### DISCUSSION TITLE
WeakPDELibrary vs ParametrizedLibrary, fulltweights vs fullweights0

### DISCUSSION THREAD

**Discussion by Jacob-Stevens-Haas** [dynamicslab/pysindy#307] [Q&A] [answered]
Hey @znicolaou, I had a question about ParametrizedLibrary.  Normally, WeakPDELibrary calculates the LHS (calc_trajectory) by multiplying values of u by `fulltweights`, i.e. "Product weights over the axes for time derivatives".  However, `ParametrizedLibrary`, when it has a weak library, then multiplies by `fullweights0`, i.e. "Product weights over the axes for pure derivative terms" (I assume this means _spatial_ derivatives).

What exactly is going on here?  Why is `ParametrizedLibrary`, essentially a thin wrapper around `GeneralizedLibrary`, calculating trajectories differently?

**Comment by znicolaou** [comment]
Hey @Jacob-Stevens-Haas ,

The issue here is a subtle one noted in the arxiv here: https://arxiv.org/abs/2301.02673, p4. The `ParameterizedLibrary` creates its feature as products of parameter features and library features. When using this with the `WeakPDELibrary` features, the product of two integrals is not the integral of the product, BUT one of the terms in the product is an integral of a constant parameter. The integral of the product of the terms is then the product of the integrals divided by the volume of the domain cell (which is contained in `fullweights0`). Rather than divide all the features by this volume, the implementation multiplies the temporal derivatives by the volume, which is equivalent. The benefit in implementing it this way is that the same code can be used for the library construction in the weak and non-weak cases, with just a small change to the temporal derivatives in the weak case.  Hope that helps--let me know if you have any questions. It could certainly be documented better...

**Reply by Jacob-Stevens-Haas** [reply]
Thanks Zach!  That makes sense in general, but I still need to think through it a bit.  Also, if it affects the lhs, is it a factor whenever a GeneralizedLibrary tensors together WeakPDELibraries?

**Reply by znicolaou** [reply, marked as answer]
Sorry for the slow response! I think currently, if a user tries to tensor together two weak libraries with a generalized library, it will calculate the product of the weak features and attempt to run the optimizer on those nonsensical features. We probably want to issue a warning at least in that case. Even in the `TensoredLibrary` class in feature_library/base.py, it does not currently check if the library is weak or is generalized library composed of weak libraries. A user could probably produce bad results if they tried complicated things naively.

**Reply by Jacob-Stevens-Haas** [reply]
Thanks Zach!  makes sense.  I'm not thinking about a code recommendation, just learning :)
